Use upstream units for mdadm

abbradar opened the pull request:
###### Motivation for this change

Fix RAID1 rebuild on each reboot.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

[WIP] because I still need to test it better. I don't have good ideas for automated testing for now.

abbradar committed:
systemd service: rename generator-packages

abbradar committed:
systemd service: add support for shutdown packages

Shutdown hooks are executed right before the shutdown, which is useful
for some applications. Among other things this is needed for mdadm hook
to run.

abbradar committed:
mdadm: use absolute paths

We also correct paths in udev rules. This is cleaner and allows for less
unexpected behaviour. We still check for self-references, however we do
it manually now and only for binaries as udev rules can have them.
Rather, we patch them out during initrd generation now.

abbradar committed:
stage-1 initrd: replace absolute paths for mdadm

We don't patch basename and readlink now too as they were added for
mdadm in 8ecd3a5e1db4.

abbradar committed:
mdadm: install systemd units

abbradar committed:
swraid service: use upstream units

This fixes a serious bug on NixOS with swraid where mdadm arrays weren't
properly stopped on shutdown. Rather than fixing the unit by adding
`Before=final.target` we completely move to upstream units, which uses
systemd shutdown hooks instead. This also drives down maintenance costs
for us.

abbradar commented:
@GrahamcOfBorg build mdadm

abbradar committed:
systemd: update revision

abbradar committed:
systemd test: add test for systemd-shutdown scripts

edolstra commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -22,10 +22,10 @@ stdenv.mkDerivation rec {
   # Also fresh patches should be cherry-picked from that tree to our current one.
   src = fetchFromGitHub {
-    owner = "NixOS";
+    owner = "abbradar";
```
Can you make a PR to the NixOS/systemd repo instead of pointing to abbradar/systemd?

abbradar replied:
I'm still testing, the first iteration of the patch didn't work. I'll do that right as my tests pass!

edolstra reviewed (commented):
Looks good, though I'm not sure if we need a separate `systemd.shutdownPackages` option. We could also just use any shutdown programs in `systemd.packages`.

abbradar commented:
@edolstra I think so too but we already have `systemd.generatorPackages`. I'm all for merging them all, what do you think?

abbradar commented:
Tested on Intel IMSM array, before the patch a reboot was practically guaranteed to start a resync. After the patch and four reboots no more resyncs. I think this is ready.

abbradar committed:
systemd service: remove generator-packages option

Use systemd.packages instead, it's less error prone and more in line with
what's expected.

abbradar commented:
This is ready with some cosmetic changes. I also removed `generator-packages` completely after thinking more, we should indeed just use `systemd.packages` everywhere.

Mic92 commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -24,8 +24,8 @@ stdenv.mkDerivation rec {
     repo = "systemd";
-    rev = "aa4c4d39d75ce52664cb28d569b1ceafda7b4c06";
-    sha256 = "1ax94gzbdwdcf3wgj7f9cabdkvn2zynnnli7gkbz4isidlpis86g";
+    rev = "6b8d75f093a0e05d9001398c81028c4b688ef801";
```
Please update to https://github.com/NixOS/systemd/commit/5fb35fbc783516e2014115c3488134a2afb8494c

Mic92 commented on pkgs/os-specific/linux/mdadm/default.nix:
```diff
@@ -8,26 +8,36 @@ stdenv.mkDerivation rec {
     sed -e 's@/lib/udev@''${out}/lib/udev@' \
         -e 's@ -Werror @ @' \
         -e 's@/usr/sbin/sendmail@${system-sendmail}@' -i Makefile
+    ls *.rules | xargs -l sed -i \
```
Why not `sed -i *.rules`?

abbradar commented:
@Mic92 All fixed, thanks!

@GrahamcOfBorg build systemd mdadm

flokli commented on nixos/tests/systemd.nix:
```diff
@@ -69,5 +77,20 @@ import ./make-test.nix {
+      # If the file system was unmounted during the shutdown the file system
+      # has a last mount time, because the file system wasn't checked.
+      $machine->fail('dumpe2fs /dev/vdb | grep -q "^Last mount time: *n/a"');
+    };
```
This seems to have been duplicated from above. Reverted in https://github.com/NixOS/nixpkgs/pull/66825.

abbradar replied:
Thanks!

flokli commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -497,11 +498,14 @@ in
       '';
     };
 
-    systemd.generator-packages = mkOption {
```
We would need a changelog entry for that, explaining people should use `systemd.packages` instead.

abbradar replied:
Won't https://github.com/NixOS/nixpkgs/pull/65376/commits/717b8b3219e0a207e16e28f828f88060b0477d0b#diff-94706aabdc6412bb34118f0202004761R224 this be enough?

flokli replied:
Yeah, it could work. I had to stare at the diff while rebasing the changes from https://github.com/NixOS/nixpkgs/pull/66856 until I figured it out.

Maybe a single line in "notable changes", explaining the new behaviour of `systemd.packages`, that it got support for system-generators (or basically all `$package/lib/systemd/system-*`) probably shouldn't hurt either.

abbradar replied:
Done in 9b30cf0cb4c0722ea1f2b85e5cb5b334d8516dc3

flokli replied:
Thanks!
